Implement boxing/unboxing during VectorAPIExpansion under an option

gita-omr opened the pull request:
- introduce new category of temps: vec_len_boxed_unknown
- build temp and whole webs as before
- wherever a temp was invalidated before assign vec_len_boxed_unknown instead
- store vectorization info for calls per node vs. per symbol reference for temps
- always vectorize nodes that are calls to vector API methods, even if input vectors need to be unboxed or result needs to be boxed
- since boxing/unboxing is not supported for all types, have an extra pass just before transformation and invalidate whole web if it contains unsupported boxing/unboxing
- transformation happens as before, except that we first check if children of non-vectorized node need to be boxed or children of vectorized node need to be unboxed, insert boxing/unboxing correspondingly
- this prototype is only enabled under -Xjit:enableVectorAPIBoxing option and the default behaviour shoulld be preserved

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.hpp:
```diff
@@ -320,20 +321,33 @@ class TR_VectorAPIExpansion : public TR::Optimization
+      TR::DataType         _elementType;
+      vapiObjType          _objectType;
+      bool                 _canVectorize;
+      bool                 _canScalarize;
```
`_canScalarize` is used nowhere in the code.

knn-k replied:
I prefer it is not added to the code until it is actually used.

gita-omr replied:
While fixing some bugs I realized we need this field, so we are using it now.

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.hpp:
```diff
@@ -552,6 +639,12 @@ class TR_VectorAPIExpansion : public TR::Optimization
    void invalidateSymRef(TR::SymbolReference *symRef);
 
+  /** \brief
+   *     Invalidared node so that it will not be vectorized or scalarized
```
`Invalidared` in this comment looks like a typo.

gita-omr replied:
done

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -44,13 +51,16 @@ TR_VectorAPIExpansion::perform()
        !TR::Compiler->om.canGenerateArraylets() &&
-       findVectorMethods(comp()))
+       findVectorMethods(comp())
+       && !comp()->isDLT() // TODO
```
Please place the `&&` at the end of the previous line, in the same way as other lines in this expression.

gita-omr replied:
done

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -348,6 +410,7 @@ TR_VectorAPIExpansion::visitNodeToBuildVectorAliases(TR::Node *node)
                               "Unexpected broadcast type in node %p\n", node);
 
                _aliasTable[methodRefNum]._objectType =(broadcastType == MODE_BROADCAST) ? Vector : Mask;
+               _nodeTable[nodeIndex]._objectType =(broadcastType == MODE_BROADCAST) ? Vector : Mask;
```
Please add a space between `=` and `(broadcastType`.

gita-omr replied:
done

knn-k replied:
Would you do the same for the previous line for `_aliasTable`, although it is existing code, please?  Thanks.

gita-omr replied:
I realized this code is not needed. Replaced with correct, simpler code. Basically, we need to determine the object type (Mask or Vector) from the first parameter of this intrinsic.

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -475,38 +571,68 @@ TR_VectorAPIExpansion::visitNodeToBuildVectorAliases(TR::Node *node)
+   else if (boxingAllowed() &&
+            node->getOpCodeValue() == TR::checkcast)
+      {
+      // allow this treetop if boxing
```
Please add a comment like "do nothing here" for readability.

gita-omr replied:
done

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -515,10 +641,22 @@ TR_VectorAPIExpansion::visitNodeToBuildVectorAliases(TR::Node *node)
+            else
+               {
+               if (_trace)
+                   traceMsg(comp(), "Invalidating10 #%d since it's used by unsupported node %p (%s)\n",
```
Perhaps you want to change "Invalidating10" to "Invalidating11" here.

gita-omr replied:
done

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -807,54 +989,85 @@ TR_VectorAPIExpansion::validateSymRef(int32_t id, int32_t i, vec_sz_t &classLeng
-         traceMsg(comp(), "%s invalidating2 class #%d due to symref #%d\n", OPT_DETAILS_VECTOR, id, i);
-      return false;
+         traceMsg(comp(), "%s invalidating13 class #%d due to symref #%d\n", OPT_DETAILS_VECTOR, id, i);
+         return false;
```
Please revert the indentation of this `return` line.

gita-omr replied:
done

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -703,6 +847,7 @@ TR_VectorAPIExpansion::getOpaqueClassBlockFromClassNode(TR::Compilation *comp, T
       else
 #endif /* defined(J9VM_OPT_JITSERVER) */
          {
+
```
Do you want an empty line here?

gita-omr replied:
done

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -912,46 +1136,60 @@ TR_VectorAPIExpansion::validateVectorAliasClasses(TR_BitVector * vectorAliasTabl
+                     break;
+                     }
                   }
+
```
Do you want an empty line here?

gita-omr replied:
fixed

gita-omr commented:
Noticed a bug..

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -2564,14 +3246,28 @@ TR::Node *TR_VectorAPIExpansion::transformNary(TR_VectorAPIExpansion *opt, TR::T
+         TR::Node* operand = operands[i];
+
+         if (operand->getOpCodeValue() == TR::aload ||
+             operand->getOpCodeValue() == TR::acall)
```
I think the code is simpler if you change this `if` statement to:

```
if (operand->getOpCodeValue() == TR::aload)
   {
   if (opt->_aliasTable[operandId]._objectType == Mask)
      {
      opCodeType = ...;
      }
   vectorizeLoadOrStore(opt, operand, opCodeType);
   }
else if (operand->getOpCodeValue() == TR::acall)
   {
   if (opt->_nodeTable[operand->getGlobalIndex()]._objectType == Mask)
      {
      opCodeType = ...;
      }
   vectorizeLoadOrStore(opt, operand, opCodeType);
   }
```

gita-omr replied:
Done.

knn-k replied:
Added another comment.

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -2564,14 +3246,28 @@ TR::Node *TR_VectorAPIExpansion::transformNary(TR_VectorAPIExpansion *opt, TR::T
+               }
+
+            vectorizeLoadOrStore(opt, operand, opCodeType);
 
```
Maybe you can remove this empty line.

gita-omr replied:
Done

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -2564,14 +3246,28 @@ TR::Node *TR_VectorAPIExpansion::transformNary(TR_VectorAPIExpansion *opt, TR::T
       for (int32_t i = 0; i < numOperands; i++)
          {
-         if (operands[i]->getOpCodeValue() == TR::aload)
+         TR::Node* operand = operands[i];
```
The majority in this file use the style `TR::Node *operand` instead of `TR::Node* operand`.

gita-omr replied:
Done

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -1186,12 +1851,29 @@ TR_VectorAPIExpansion::vectorizeLoadOrStore(TR_VectorAPIExpansion *opt, TR::Node
-      TR::Node::recreate(node, TR::ILOpCode::createVectorOpCode(opCodeType.isVector() ? TR::vload : TR::mload, opCodeType));
+      {
+      TR_ASSERT_FATAL(!node->getOpCode().isStore(), "Should be a load node");
+      node = TR::Node::create(node, opcode, 0);
```
You assign the return value of `TR::Node::recreate()` to `node` here, while you don't do that at L.1864.  Is that OK?
I searched for the definition of `TR::Node::recreate()` with three arguments, but I couldn't find it.

gita-omr replied:
This is by design. In one case, we modify the node in place and in another (new) case we create a new node.

gita-omr commented:
Fixed some bugs and addressed second round of comments. openjdk sanity passes (with and without the option). However, I had to disable boxing of Mask objects due to a failure in that bucket. Debugging now.

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -2564,14 +3267,31 @@ TR::Node *TR_VectorAPIExpansion::transformNary(TR_VectorAPIExpansion *opt, TR::T
+            {
+            TR::DataType opCodeType = vectorType;
+            TR::SymbolReference *operandSymRef = operand->getSymbolReference();
+            int32_t operandId = operandSymRef->getReferenceNumber();
```
You don't need `operandId` and `operandSymRef` in this path, do you?

gita-omr replied:
Fixed

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -962,45 +1205,362 @@ TR_VectorAPIExpansion::validateVectorAliasClasses(TR_BitVector * vectorAliasTabl
+      {
+      _classByte128Vector = comp()->fej9()->getClassFromSignature("jdk/incubator/vector/Byte128Vector", 34, owningMethod, true);
+      TR_ASSERT_FATAL(_classByte128Vector, "Could not create Vector class from signature");
+
```
Do you want an empty line here?
The same comment for the line after `TR_ASSERT_FATAL(_classByte128Mask, "Could not create Mask class from signature");`

gita-omr replied:
Fixed

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -729,6 +879,42 @@ TR_VectorAPIExpansion::getDataTypeFromClassNode(TR::Compilation *comp, TR::Node
+         j9class = vm->byteArrayClass;
+         break;
+      case TR::Int16:
+         j9class = vm->charArrayClass;
```
Is there any reason for using `charArrayClass` instead of `shortArrayClass`?
All other cases with `TR::Int` in this switch statement use array classes of signed integer types.

gita-omr replied:
You are right, I think we need `shortArrayClass`. Fixed.

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -729,6 +879,42 @@ TR_VectorAPIExpansion::getDataTypeFromClassNode(TR::Compilation *comp, TR::Node
+      }
+
+   return fej9->convertClassPtrToClassOffset(j9class);
+
```
Do you want an empty line here?

gita-omr replied:
Fixed

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -962,45 +1205,362 @@ TR_VectorAPIExpansion::validateVectorAliasClasses(TR_BitVector * vectorAliasTabl
+   // Generate vector store to the payload array
+
+   TR::VectorLength vectorLength = OMR::DataType::bitsToVectorLength(bitsLength);
+   TR::DataType opCodeType = TR::DataType::createVectorType(elementType, vectorLength);
+
+   if (objectType == Mask)
+       opCodeType = TR::DataType::createMaskType(elementType, vectorLength);
```
Do you need to call `createVectorType()` when `objectType == Mask`?
If no, how about changing these lines to:
```
TR::DataType opCodeType = (objectType == Vector) ?
  TR::DataType::createVectorType(elementType, vectorLength) :
  TR::DataType::createMaskType(elementType, vectorLength);
```

gita-omr replied:
This code was rewritten.

gita-omr replied:
Sorry, similar code was rewritten somewhere else. Done.

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -1128,13 +1750,82 @@ TR_VectorAPIExpansion::expandVectorAPI()
+                  bool vectorized = false;
+
+                  {
+                  TR::DataType elementType;
```
Please use a different variable name at least for `elementType` because another `elementType` is declared in the outer scope.

gita-omr replied:
Done

knn-k commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -1027,19 +1604,49 @@ TR_VectorAPIExpansion::expandVectorAPI()
          continue;
 
-      if (node->chkStoredValueIsIrrelevant())
+      TR::DataType elementType;
+      int32_t bitsLength;
+      vapiObjType objectType;
+
```
Can you limit the scope of these variables or use other names?
Variables with the same names are declared and used in the `if (opCodeValue == TR::astore) { ... }` block below.

gita-omr replied:
Done

gita-omr commented:
Added improvements and addressed latest comments except for the doxygen comments.

gita-omr commented:
Squashed and removed WIP.

gita-omr commented:
Addressed latest comment.

knn-k commented:
Jenkins test sanity alinux64,plinux,xlinux,zlinux jdk21

gita-omr commented:
s390 failures seem to be unrelated.

knn-k commented:
Jenkins test sanity alinux64,plinux,xlinux,zlinux jdk21

knn-k commented:
A test job on xlinux failed with the message below:
https://openj9-jenkins.osuosl.org/job/Test_openjdk21_j9_sanity.functional_x86-64_linux_Personal_testList_1/189/consoleText

```
[2024-12-06T02:27:21.089Z] Boooo - there are rogue processes kicking about
[2024-12-06T02:27:21.089Z] Issuing a kill to all processes shown above
```

knn-k commented:
Jenkins test sanity.functional xlinux jdk21

knn-k commented:
Test failures on plinux and zlinux are related to CRIU.  They don't seem to be caused by this PR.  Waiting for xlinux results.

**plinux**:
cmdLineTester_criu_jitserverAcrossCheckpoint_0
https://openj9-jenkins.osuosl.org/job/Test_openjdk21_j9_sanity.functional_ppc64le_linux_Personal_testList_0/127/consoleText

**zlinux**:
cmdLineTester_criu_jitPostRestore_3
https://openj9-jenkins.osuosl.org/job/Test_openjdk21_j9_sanity.functional_s390x_linux_Personal_testList_1/269/consoleText

mpirvu commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -729,6 +895,45 @@ TR_VectorAPIExpansion::getDataTypeFromClassNode(TR::Compilation *comp, TR::Node
+   switch (type)
+      {
+      case TR::Float:
+         j9class = vm->floatArrayClass;
```
FYI: Accessing J9JavaVM directly in the optimizer will break JITServer functionality. Typically a frontend function needs to be created to fetch the appropriate field and that function needs to be overridden at the server frontend. PR https://github.com/eclipse-openj9/openj9/pull/20592 shows an example

mpirvu commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -962,45 +1220,422 @@ TR_VectorAPIExpansion::validateVectorAliasClasses(TR_BitVector * vectorAliasTabl
+   if (_trace)
+      traceMsg(comp(), "Boxed: %dth child of node %p into %p\n", i, node, newObject);
+
+   if (TR::Options::getVerboseOption(TR_VerboseVectorAPI))
```
I would prefer the optimizer to use compilation logs instead of vlog. Vlog is typically used for runtime. Verbose options are global and they are not passed to the JITServer If the method is compiled remotely, these messages will not appear.
